sql: rewrite UDF references in views to be by ID

rafiss opened the pull request:
### sql: rewrite UDF references in views to be by ID

This allows for renaming UDFs that are referenced by views. We use the
same logic that we use for rewriting DEFAULT or computed column
expressions.

To get this to work, we need to be able to type-check the exprssions in
the view query, and that required resolving column references by using
the optimizer's plan dependencies.

### sql: improve column name resolution when resolving UDFs in views

This patch makes it so we can handle a view query that has two source
tables that have identical column tables. If the source tables also have
identical names, then we fall back to not attempting to resolve the
column type.

fixes #146123
informs https://github.com/cockroachdb/cockroach/issues/87699

Release note: None

rafiss committed:
sql: rewrite UDF references in views to be by ID

This allows for renaming UDFs that are referenced by views. We use the
same logic that we use for rewriting DEFAULT or computed column
expressions.

To get this to work, we need to be able to type-check the exprssions in
the view query, and that required resolving column references by using
the optimizer's plan dependencies.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146475)
<!-- Reviewable:end -->

rafiss committed:
sql: improve column name resolution when resolving UDFs in views

This patch makes it so we can handle a view query that has two source
tables that have identical column tables. If the source tables also have
identical names, then we fall back to not attempting to resolve the
column type.

Release note: None

rafiss committed:
sql: cleanup multiStmt parameter for helper functions

This was always false, so the parameter can be removed.

Release note: None

DrewKimball reviewed (commented):
> Reviewed 6 of 8 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 6 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfv7RV3-ARy_NaXzLV:-OUfv7RW3MMErmpYKYIK:b-ubhk87):*
nit: typo in `exprssions`
___
*[`-- commits` line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfu2WjExvXsmc6c4Xi:-OUfu2WjExvXsmc6c4Xj:bxiip0o):*
nit: `identical column tables` -> `identical column names`
___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUfxaIG3U3oHKh2OxlN:byoth52) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
> ```Go
> 				if !hasDuplicates {
> 					for _, dep := range deps {
> 						if columnItem.TableName == nil || dep.desc.GetName() != columnItem.TableName.Object() {
> ```

This might be ok for now, but probably deserves a TODO/issue. IUUC, it means we can't use schema/database to disambiguate data source names, is that right? Note that the optbuilder makes sure to format the AST so that data source names are fully qualified.

I'm also curious what happens if a table is given an alias, so that the column reference is like `alias.colName`. Would that prevent this descriptor check from finding the column? What about a case like this?
```
CREATE TABLE t1 (x INT, y TEXT);
CREATE TABLE t2 (x TEXT, y INT);
INSERT INTO t1 VALUES (1, '1');
INSERT INTO t2 VALUES ('100', 100);

CREATE FUNCTION f(a INT, b TEXT) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b::INT $$;
CREATE FUNCTION f(a TEXT, b INT) RETURNS INT LANGUAGE SQL AS $$ SELECT -1 * (a::INT + b) $$;

CREATE VIEW v AS SELECT f(t1.x, t2.x) FROM t1 AS t2, t2 AS t1;
SELECT * FROM v;

DROP VIEW v;
DROP FUNCTION f(INT, TEXT);
DROP FUNCTION f(TEXT, INT);
DROP TABLE t1;
DROP TABLE t2;
```
___
*[`pkg/sql/sem/tree/constant.go` line 411 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUftfno8I1YeoU2fcer:-OUftfno8I1YeoU2fces:bqnzqq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b4d4c86920b652ed0b4ec1bb8af5c149b661d5/pkg/sql/sem/tree/constant.go#L411)):*
> ```Go
> 		}
> 		o, err := IntToOid(MustBeDInt(d))
> 		return NewDOidWithType(o, typ), err
> ```

What prompted this change? It makes sense to me, but I'm wondering what types of queries it will impact.
___
*[`pkg/sql/create_function.go` line 601 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfsvUn2-FTLR7ox3Md:-OUfsvUn2-FTLR7ox3Me:b-8yagke) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d13703e8a519dc175c09a93f5ea11c058366aca/pkg/sql/create_function.go#L601)):*
> ```Go
> 				return err
> 			}
> 			body, err = serializeUserDefinedFunctionsLang(
> ```

Will this allow renaming routines referenced by other routines?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> Reviewed 6 of 8 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 7 of 7 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`-- commits` line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUg4BMs2QQqbRev4XyY:-OUg4BMs2QQqbRev4XyZ:bc4beps):*
Is this because different function overloads have different OIDs? Do we need some other OID-like representation that doesn't specify an overload?
___
*[`pkg/sql/alter_function.go` line 213 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUg-DrwCvUpD3lU8VOl:-OUg-DrwCvUpD3lU8VOm:b-giuckc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d13703e8a519dc175c09a93f5ea11c058366aca/pkg/sql/alter_function.go#L213)):*
> ```Go
> 	}
> 
> 	// Disallow renaming if this rename operation will break other UDF's or views
> ```

nit: comment needs updating
___
*[`pkg/sql/alter_function.go` line 537 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUg-caeE-OdK9aM4JAT:-OUg-caeE-OdK9aM4JAU:b-fujgaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d13703e8a519dc175c09a93f5ea11c058366aca/pkg/sql/alter_function.go#L537)):*
> ```Go
> 		}
> 		switch desc.(type) {
> 		case catalog.FunctionDescriptor:
> ```

Why do function references still disallow alter, now that we're calling `serializeUserDefinedFunctionsLang` in `setFuncOptions`? Is that not enough to fix #83233 and #87699?
___
*[`pkg/sql/create_function.go` line 601 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfyNycCePolCYLItv0:-OUfyNycCePolCYLItv1:b-gvtjrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d13703e8a519dc175c09a93f5ea11c058366aca/pkg/sql/create_function.go#L601)):*
> ```Go
> 				return err
> 			}
> 			body, err = serializeUserDefinedFunctionsLang(
> ```

nit: a matching comment would be nice
___
*[`pkg/sql/create_view.go` line 759 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfvQFNElU4qbg0SKGN:-OUfvQFNElU4qbg0SKGO:bpf8x0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d13703e8a519dc175c09a93f5ea11c058366aca/pkg/sql/create_view.go#L759)):*
> ```Go
> 		}
> 		// semaCtx may be nil if this is a virtual view being created at init time.
> 		if semaCtx == nil {
> ```

Should this check be hoisted to the top of `serializeUserDefinedFunctionsLang`?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUgJGZf8OlfLvtdqB0K:b-djzx07) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This might be ok for now, but probably deserves a TODO/issue. IUUC, it means we can't use schema/database to disambiguate data source names, is that right? Note that the optbuilder makes sure to format the AST so that data source names are fully qualified.

I'm also curious what happens if a table is given an alias, so that the column reference is like `alias.colName`. Would that prevent this descriptor check from finding the column? What about a case like this?
```
CREATE TABLE t1 (x INT, y TEXT);
CREATE TABLE t2 (x TEXT, y INT);
INSERT INTO t1 VALUES (1, '1');
INSERT INTO t2 VALUES ('100', 100);

CREATE FUNCTION f(a INT, b TEXT) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b::INT $$;
CREATE FUNCTION f(a TEXT, b INT) RETURNS INT LANGUAGE SQL AS $$ SELECT -1 * (a::INT + b) $$;

CREATE VIEW v AS SELECT f(t1.x, t2.x) FROM t1 AS t2, t2 AS t1;
SELECT * FROM v;

DROP VIEW v;
DROP FUNCTION f(INT, TEXT);
DROP FUNCTION f(TEXT, INT);
DROP TABLE t1;
DROP TABLE t2;
```
</blockquote></details>

as you predict, that example does have an issue. the view actually does get created, but during execution, it hits:
```
root@localhost:26257/defaultdb> SELECT * FROM v;;
ERROR: unknown signature: public.f(string, int)
SQLSTATE: 42883
HINT: No function matches the given name and argument types. You might need to add explicit type casts.
```

I also realized another issue with this function -- it can't handle stars at all:
```
root@localhost:26257/defaultdb> CREATE VIEW v3 AS SELECT count(*) FROM t1;
ERROR: count(): cannot use "*" in this context
SQLSTATE: 42601
```

I think that's because the `ReplaceColumnVars` function doesn't handle `*`.

so that means we probably should use a different approach entirely. Type checking the view query here seems fraught. I wonder if we could do the function reference replacement in the optimizer instead, since that's the place that already type checks the view query.

The other option would be to figure out what we need in order to do type checking properly here, (meaning expand `*`s and column references) without using the hacky approach this commit has right now. 

Do you have thoughts on how to tackle either of those?
___
*[`pkg/sql/sem/tree/constant.go` line 411 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUftfno8I1YeoU2fcer:-OUgJpbF2kON6dTS4dV3:b1jsgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b4d4c86920b652ed0b4ec1bb8af5c149b661d5/pkg/sql/sem/tree/constant.go#L411)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What prompted this change? It makes sense to me, but I'm wondering what types of queries it will impact.
</blockquote></details>

I needed this because the view replace helper function would parse the query string and then re-format it into a string later. during the type checking step, the sequence-replaced expression (e.g. `nextval(106::regclass)`) was getting formatted as `nextval(106::oid)` because we were losing the type of the dOID during type-checking.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUgTxo6Dzc2N-_cQoxo:b-fb393w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

as you predict, that example does have an issue. the view actually does get created, but during execution, it hits:
```
root@localhost:26257/defaultdb> SELECT * FROM v;;
ERROR: unknown signature: public.f(string, int)
SQLSTATE: 42883
HINT: No function matches the given name and argument types. You might need to add explicit type casts.
```

I also realized another issue with this function -- it can't handle stars at all:
```
root@localhost:26257/defaultdb> CREATE VIEW v3 AS SELECT count(*) FROM t1;
ERROR: count(): cannot use "*" in this context
SQLSTATE: 42601
```

I think that's because the `ReplaceColumnVars` function doesn't handle `*`.

so that means we probably should use a different approach entirely. Type checking the view query here seems fraught. I wonder if we could do the function reference replacement in the optimizer instead, since that's the place that already type checks the view query.

The other option would be to figure out what we need in order to do type checking properly here, (meaning expand `*`s and column references) without using the hacky approach this commit has right now. 

Do you have thoughts on how to tackle either of those?
</blockquote></details>

Do we need to pick the correct overload for the function call? Could we somehow use `types.Any` to pick an arbitrary overload, and then rewrite it back to a function name using `tree.OidToBuiltinName` before execution?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
Now that we store the function in the UDF by ID, does this PR need to handle OID rewrite during a restore (`sql/catalog/rewrite/rewrite.go`)? 

EDIT: I was playing around with a few scenarios before leaving this comment. There might be a broader issue outside of this PR. I logged this issue: https://github.com/cockroachdb/cockroach/issues/149782

> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-:-OUjPswd1oL4dY2RpWmx:bumd05f)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUk1V1D33Rc2mEP1Oih:b-xw9l23) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
> Do we need to pick the correct overload for the function call? Could we somehow use `types.Any` to pick an arbitrary overload, and then rewrite it back to a function name using `tree.OidToBuiltinName` before execution?

That seems like it could work conceivably, but introduces some risk if a code path appears later that forgets to rewrite before execution. I can investigate this though -- do you happen to already know where the code for executing the view query resides currently, and where the rewrite should happen?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUk8vtE066WSGLzSE2x:btvkguq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> Do we need to pick the correct overload for the function call? Could we somehow use `types.Any` to pick an arbitrary overload, and then rewrite it back to a function name using `tree.OidToBuiltinName` before execution?

That seems like it could work conceivably, but introduces some risk if a code path appears later that forgets to rewrite before execution. I can investigate this though -- do you happen to already know where the code for executing the view query resides currently, and where the rewrite should happen?
</blockquote></details>

Here's where we handle adding the view query: https://github.com/cockroachdb/cockroach/blob/5128b4772335eb67ba0fb2e6395d81cb4b815d68/pkg/sql/opt/optbuilder/select.go#L305


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUk9N-J-FkcDxrITbD8:b58fl8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Here's where we handle adding the view query: https://github.com/cockroachdb/cockroach/blob/5128b4772335eb67ba0fb2e6395d81cb4b815d68/pkg/sql/opt/optbuilder/select.go#L305
</blockquote></details>

(Though I guess one problem with this approach is that it wouldn't allow dropping specific overloads of a UDF.)


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146475)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/create_view.go` line 783 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146475#-OUfxaIG3U3oHKh2OxlM:-OUk9zAY9NIvHiUfT_Hs:b-jet89p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f8b04cdb3235d28ecd4646ffbd8f7ee4cde03/pkg/sql/create_view.go#L783)):*
if we're going to be doing rewrites in the optimizer anyway, then it feels cleaner to me to pursue the idea i mentioned above to replace the name with the by-ID reference in the optimizer at the time of creating the view: 

> I wonder if we could do the function reference replacement in the optimizer instead, since that's the place that already type checks the view query.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
>I wonder if we could do the function reference replacement in the optimizer instead, since that's the place that already type checks the view query.

I think this is a good idea. We already rewrite relation references to be fully qualified, so there's some precedent to rewriting the AST in the optimizer.

rafiss commented:
I'm closing this since it's taking more time than expected, and we have decided to make this by-ID reference project not block the 25.3 release. see https://github.com/cockroachdb/cockroach/issues/146123#issuecomment-3071007333

I attempted to do the rewrite in the optimizer (https://github.com/cockroachdb/cockroach/pull/149863), and that also ran into some trouble.
